llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -2727,105 +2727,161 @@ Instruction *InstCombinerImpl::matchBSwapOrBitReverse(Instruction &I,
       return nullptr;
+    };
 
-    // For non-constant cases we don't support non-pow2 shift masks.
-    // TODO: Is it worth matching urem as well?
-    if (!isPowerOf2_32(Width))
+    Value *ShAmt = matchShiftAmount(ShAmt0, ShAmt1, Width);
+    if (!ShAmt) {
+      ShAmt = matchShiftAmount(ShAmt1, ShAmt0, Width);
+      IsFshl = false; // Sub on SHL.
+    }
+    if (!ShAmt)
       return nullptr;
 
-    // The shift amount may be masked with negation:
-    // (shl ShVal, (X & (Width - 1))) | (lshr ShVal, ((-X) & (Width - 1)))
-    Value *X;
-    unsigned Mask = Width - 1;
-    if (match(L, m_And(m_Value(X), m_SpecificInt(Mask))) &&
-        match(R, m_And(m_Neg(m_Specific(X)), m_SpecificInt(Mask))))
-      return X;
+    FShiftArgs = {ShVal0, ShVal1, ShAmt};
 
-    // Similar to above, but the shift amount may be extended after masking,
-    // so return the extended value as the parameter for the intrinsic.
-    if (match(L, m_ZExt(m_And(m_Value(X), m_SpecificInt(Mask)))) &&
-        match(R, m_And(m_Neg(m_ZExt(m_And(m_Specific(X), m_SpecificInt(Mask)))),
-                       m_SpecificInt(Mask))))
-      return L;
+  } else if (isa<ZExtInst>(Or0) || isa<ZExtInst>(Or1)) {
+    // If there are two 'or' instructions concat variables in opposite order,
+    // the latter one can be safely convert to fshl.
+    //
+    // LowHigh = or (shl (zext Low), Width - ZextHighShlAmt), (zext High)
+    // HighLow = or (shl (zext High), ZextHighShlAmt), (zext Low)
+    // ->
+    // HighLow = fshl LowHigh, LowHigh, ZextHighShlAmt
+    if (!isa<ZExtInst>(Or1))
+      std::swap(Or0, Or1);
+
+    Value *High, *ZextHigh, *Low;
+    const APInt *ZextHighShlAmt;
+    if (!match(Or0,
+               m_OneUse(m_Shl(m_Value(ZextHigh), m_APInt(ZextHighShlAmt)))))
+      return nullptr;
 
-    if (match(L, m_ZExt(m_And(m_Value(X), m_SpecificInt(Mask)))) &&
-        match(R, m_ZExt(m_And(m_Neg(m_Specific(X)), m_SpecificInt(Mask)))))
-      return L;
+    if (!match(Or1, m_ZExt(m_Value(Low))) ||
+        !match(ZextHigh, m_ZExt(m_Value(High))))
+      return nullptr;
 
-    return nullptr;
-  };
+    unsigned HighSize = High->getType()->getScalarSizeInBits();
+    unsigned LowSize = Low->getType()->getScalarSizeInBits();
+    if (*ZextHighShlAmt != LowSize || HighSize + LowSize != Width)
+      return nullptr;
 
-  Value *ShAmt = matchShiftAmount(ShAmt0, ShAmt1, Width);
-  bool IsFshl = true; // Sub on LSHR.
-  if (!ShAmt) {
-    ShAmt = matchShiftAmount(ShAmt1, ShAmt0, Width);
-    IsFshl = false; // Sub on SHL.
+    for (User *U : ZextHigh->users()) {
+      Value *X, *Y;
+      if (!match(U, m_Or(m_Value(X), m_Value(Y))))
+        continue;
+
+      if (!isa<ZExtInst>(Y))
+        std::swap(X, Y);
+
+      if (match(X, m_Shl(m_Specific(Or1), m_SpecificInt(HighSize))) &&
+          match(Y, m_Specific(ZextHigh)) && DT.dominates(U, &Or)) {
+        FShiftArgs = {U, U, ConstantInt::get(Or0->getType(), *ZextHighShlAmt)};
+        break;
+      }
+    }
```

Comment:
As a general rule, we don't do combines that require looking for partner instructions via use-list iteration. This is really the type of fold that makes more sense in a DAG context.

If this where a high-ticket fold it might be different, but generally I'm opposed to this type of transform in InstCombine.

Reply:
Given IR like this.
```
  %zext.x = zext i8 %x to i32
  %slx = shl nuw i32 %zext.x, 24
  %zext.y = zext i24 %y to i32
  %xy = or i32 %zext.y, %slx
  store i32 %xy, ptr %addr, align 4

  %sly = shl nuw i32 %zext.y, 8
  %yx = or i32 %zext.x, %sly
  %and.i3 = and i32 %yx, 16711935
  %shl1.i = shl nuw i32 %and.i3, 8
  %and2.i = and i32 %yx, -16711936
  %shr3.i = lshr i32 %and2.i, 8
  %or4.i = or i32 %shl1.i, %shr3.i
  %and5.i = and i32 %or4.i, 252645135
  %shl6.i = shl nuw i32 %and5.i, 4
  %and7.i = and i32 %or4.i, -252645136
  %shr8.i = lshr i32 %and7.i, 4
  %or9.i = or i32 %shl6.i, %shr8.i
  %and10.i = and i32 %or9.i, 858993459
  %shl11.i = shl nuw i32 %and10.i, 2
  %and12.i = and i32 %or9.i, -858993460
  %shr13.i = lshr i32 %and12.i, 2
  %or14.i = or i32 %shl11.i, %shr13.i
  %and15.i = and i32 %or14.i, 1431655765
  %shl16.i = shl nuw i32 %and15.i, 1
  %and17.i = and i32 %or14.i, -1431655766
  %shr18.i = lshr i32 %and17.i, 1
  %or19.i = or i32 %shl16.i, %shr18.i
```
Apparently it can be optimized to
```
  %zext.x = zext i8 %x to i32
  %slx = shl nuw i32 %zext.x, 24
  %zext.y = zext i24 %y to i32
  %xy = or i32 %zext.y, %slx
  store i32 %xy, ptr %addr, align 4

  %res = call i32 @llvm.bitreverse.i32(i32 %xy)
```
This requires first optimize
```
  %sly = shl nuw i32 %zext.y, 8
  %yx = or i32 %zext.x, %sly
```
to 
`  fshl %xy, %xy, 8
`
Then teach InstCombine to optimize the rest to bitreverse.
If we do this funnel shift optimization in DAG combiner. There's no chance to optimize the rest to bitreverse.

Reply:
This is a wrong example.
Please look at this correct example: https://github.com/llvm/llvm-project/pull/68502#issuecomment-1807132251